Ensure that concurrent version updates play well together

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/21300
## Description

This PR adds locking around the assignment of a journal version. This ensures that there are no intermittent `INSERT`s into the journals table between determining the latest version of a journable and actually adding the next version.
### PostgreSQL

For PSQL we can use a `SHARE ROW EXCLUSIVE` lock on the journals table. While held, this locks prevents:
- concurrent `INSERT`/`DELETE`/`UPDATE` statements
- concurrent acquisition of a `SHARE ROW EXCLUSIVE` lock (i.e. it is self-exclusive)

The lock is released upon `COMMIT` or `ROLLBACK` of the surrounding transaction, which is neccessary, because otherwise a concurrent transaction would not be able to see the changes of the lock bearer and therefore still try to insert an outdated value for `version`.
### MySQL

Opposed to PostgreSQL, in MySQL there are no locks that are released at the end of a transaction. Table locks of all kinds implicitly commit running transactions (and starting a transaction releases all table locks).
Advisory locks (the ones without meaning for the DBMS, but usable by applications) at least don't commit your transactions, but you must release them independently of your transactions.

This means we can use advisory locks to implement self exclusiveness, but we have to release it after our transaction finished (`COMMIT` or `ROLLBACK`) manually. Since the code starting the transaction should not know about the implementation details of the code relying on the advisory lock (those are oftentimes different parts of our application), I extracted the release-on-transaction-end behaviour into a separate gem called [transactional_lock](https://github.com/finnlabs/transactional_lock).

Note: Please see the transactional_lock gem as a part of this PR and include it in your review.

After review, we might want to publish it like any regular gem on rubygems.org and use it from there instead of using a git reference.

NobodysNightmare committed:
Ensure that concurrent version updates play well together

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-lock-version-updates.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-lock-version-updates.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
Fix saving of journal in association

When moving the code over here I replaced a
journable.journals.select(&:new_record?).each(&:save!)
with
journal.save!

I did this because we know that only the latest journal will be unsaved.
However, it seems that we actually rely on a side-effect of journable.journals being iterated
(journable.journals.to_a works as well).

I have no idea whats going on, but I made the unobvious part more obvious...

ulferts commented:
The solution for Postgres seems to work. But I don't like it. This will lock the entire table for reading untill the transaction is through resulting in poor performance. And of course the lack of a solution for MySQL is a serious flaw.

Unfortunately, I don't have a better one. Here are some alternatives that I came across when searching the web for a solution to the problem:

1) Have a separate table just for the versions (Columns are made up from the keys in the unique constraint: journalizable_type, journalizable_id, version) where one can then lock the individual row. This will result in a partial lock for the journalizable's journals that are going to be updated. But it would require us to maintain a separate table just for locking.
2) Similar to the one before, one can rely on sequences.
3) Use `INSERT … ON DUPLICATE KEY UPDATE` (MySQL) or `INSERT ... ON CONFLICT {UPDATE | IGNORE}` (Postgresql). I am not sure about the concurrency guarantees that exist. Additionally, it is only available for Postgresql newer or equal to 9.5. Additionally, I don't know the extend to which one can specify the conditions for a conflict.
4) Specify functions to take care of that for you  http://stackoverflow.com/questions/6369152/postgresql-auto-increment-based-on-multi-column-unique-constraint
5) Get rid of the `versions` column and calculate the version when querying based on the `created_at` column. That however would require to change a considerable amount of the interactions with journals and it will only come at no additional performance impact if we always query for all journals of a journalizable.

So, let the debate begin.

ulferts commented:
Thinking about this in the morning brought me to the conclusion that all of my above proposals will lead to deadlocks in certain situations.

Because updates to work packages are inherited up (estimated time, remaining hours) the parent-child relationship as well as down (target version) additionally to our reliance on the `lft` and `rgt` values we need for the nested set implementation, concurrent updates on two tickets within a work package tree can cause deadlocks if we don't lock considerably more than a single row. Taking the other possible relationships (blocks, follows, ...) into account we either need to have a very elaborate locking mechanism or we need to lock the whole table.

I'd take locking the whole table for now.

Therefore we need to come up with a good solution for mysql. The postgresql solution is good IMO. 

A solution not involving explicit locking might be beneficial as well. A solution that has been proposed is using the `SERIALIZABLE` level for a transaction (http://blog.2ndquadrant.com/postgresql-anti-patterns-read-modify-write-cycles/). Than we would have to deal with failing transactions but we have to deal with those in any case.

NobodysNightmare committed:
Use an advisory lock that is released on COMMIT/ROLLBACK

for MySQL this seems to be the only available workaround that is not destroying running transactions.

Note that TransactionalLock is a self-developed gem, that will emulate PostgreSQL style advisory locks, that are automatically released at transaction end.

NobodysNightmare committed:
Enforce default isolation level of READ COMMITTED

This sets the session (== ActiveRecord connection) wide default transaction isolation level.

Specific reason: For advisory based locking over journals in MySQL to work correctly,
we need to ensure that the SELECT that determines the maximum version reads from the most recent committed data.
However, by default MySQL would use REPEATABLE READ, which snapshots the database at the first SELECT of a transaction.

Generic reason: Until now we did not expect any specific isolation level,
this might be bad as MySQL and PostgreSQL already use different defaults. It is also possible to reconfigure a DBMS for a different default, so we would either need to require our customers to configure a specified isolation level OR we set one for our own sessions... Now we know for sure that our transactions are using READ COMMITTED unless specified otherwise on the transaction.

NobodysNightmare commented:
I declare this as ready for review, since it is now working for MySQL and PostgreSQL.

I'd really like to add specs, if anyone has a good idea how to do this without building a big bad flickerer contact me...

edit: maybe not so ready...
edit2: now it looks better...

NobodysNightmare committed:
Resolve stackoverflow

new_connection -> our patch -> determine adapter type -> ActiveRecord::Base.connection -> new_connection

ulferts commented on Gemfile:
```diff
@@ -98,6 +98,8 @@ gem 'gon', '~> 4.0'
 # don't require by default, instead load on-demand when actually configured
 gem 'airbrake', '~> 4.1.0', require: false
 
+gem 'transactional_lock', git: 'https://github.com/finnlabs/transactional_lock.git', branch: 'master'
```
I'd rather see this within the 

```
platforms :mri, :mingw, :x64_mingw do
  group :mysql2 do
    gem 'mysql2', '~> 0.3.11'
  end
  ...
end
```

block

NobodysNightmare replied:
Wouldn't that mean I am only installing it for those platforms?
What about the `jruby` platform below?

I don't quite see how this gem should be specific to a certain platform...

NobodysNightmare replied:
Okay, after some back and forth, I've now left the gem where it was, because it does not depend on any specific database backend itself, there is no gem dependency.

It will (as of today) only work correctly for a MySQL backend, but this is no design choice, but rather a limitation of the implementation (of the gem).

ulferts commented on app/models/journal.rb:
```diff
@@ -56,6 +56,29 @@ class Journal < ActiveRecord::Base
+        # MySQL is very weak when combining transactions and locks. Using an emulation layer to
+        # automatically release an advisory lock at the end of the transaction
+        # FIXME: still creates duplicates... because MySQL defaults to READ REPEATABLE isolation
+        # we need READ COMMITED, which is also the default for PostgreSQL
```
The FIXME is no longer up to date right? https://github.com/opf/openproject/pull/3482/files#diff-61fefd1ac4f13f6ec6cebc79a052988dR36

ulferts commented on config/initializers/transactional_lock.rb:
```diff
@@ -0,0 +1,34 @@
+  config.after_initialize do
+    TransactionalLock.initialize
+  end
+end
```
`config/initializers` is meant for pieces of code that runs during the `initialize` phase of the rails boot up. I'd simply put this piece of code inside `config/application.rb`

NobodysNightmare replied:
Not really sure about that. I thought it was in line with files like: https://github.com/opf/openproject/blob/dev/config/initializers/airbrake.rb

Also the `application.rb` states:

> ```
> # Application configuration should go into files in config/initializers
> ```

ulferts commented:
That's all from my side for now @NobodysNightmare

NobodysNightmare commented:
@ulferts I applied your initial review feedback on the gem here: https://github.com/finnlabs/transactional_lock/pull/2

after you are satisfied with that I will update the reference in the `Gemfile` to the merged version.

NobodysNightmare committed:
Remove outdated comments

NobodysNightmare committed:
Increase default lock timeout
